### PR TITLE
feat: Implemented angular calendar view for appointments

### PR DESCRIPTION
**Pull request by theFaustus**

### CODE REVIEW

File: med-assist-backend/src/main/java/inc/evil/medassist/appointment/facade/AppointmentFacadeImpl.java [hunk 1]
```diff
@@ -71,15 +66,31 @@ public AppointmentResponse update(String id, UpsertAppointmentRequest request) {
...
+    }
+
+    private Appointment toAppointment(final UpsertAppointmentRequest request) {
+        Appointment appointmentToCreate = Appointment.builder()
+                .appointmentDate(request.getStartDate() != null ? request.getStartDate().toLocalDate() : null)
+                .startTime(request.getStartDate() != null ? request.getStartDate().toLocalTime() : null)
+                .endTime(request.getEndDate() != null ? request.getEndDate().toLocalTime() : null)
                 .operation(request.getOperation())
                 .doctor(request.getDoctorId() != null ? doctorService.findById(request.getDoctorId()) : null)
-                .patient(request.getPatientId() != null ? patientService.findById(request.getPatientId()) : null)
```
Comment (anrosca) [inline med-assist-backend/src/main/java/inc/evil/medassist/appointment/facade/AppointmentFacadeImpl.java:81]: Why don't we add a method like this:
```java
private Patient findOrCreatePatient(final UpsertAppointmentRequest request) {
        if(request.isExistingPatient()){
           return request.getPatientId() != null ? patientService.findById(request.getPatientId()) : null;
        } else {
           return patientService.create(request.getPatientRequest().toPatient());
        }
}
```

and then while building the `appointmentToCreate` object we can have something like:

```java
...
.doctor(request.getDoctorId() != null ? doctorService.findById(request.getDoctorId()) : null)
.patient(findOrCreatePatient(request))
...
```

Comment (theFaustus) [reply med-assist-backend/src/main/java/inc/evil/medassist/appointment/facade/AppointmentFacadeImpl.java:81]: agree, great idea

File: med-assist-backend/src/test/java/inc/evil/medassist/common/validation/AtLeastOneNotNullConstraintValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AtLeastOneNotNullConstraintValidatorTest {
+    @Spy
+    private AtLeastOneNotNullConstraintValidator atLeastOneNotNullConstraintValidator;
```
Comment (anrosca) [inline med-assist-backend/src/test/java/inc/evil/medassist/common/validation/AtLeastOneNotNullConstraintValidatorTest.java:28]: This test is kind of too tied to the actual implementation. We can try to verify the behavior instead, something like this:
```java

public class AtLeastOneNotNullConstraintValidatorTest {
    private Validator validator;

    @BeforeEach
    private void setUp() {
        ValidatorFactory factory = Validation.buildDefaultValidatorFactory();
        validator = factory.getValidator();
    }
    
    @Test
    public void isValid_whenOneFieldIsNotNull_isTrue() {
        ObjectToValidate object = new ObjectToValidate("field1", null);

        Set<ConstraintViolation<ObjectToValidate>> actualViolations = validator.validate(object);

        assertThat(actualViolations).isEmpty();
    }

    @Test
    public void isValid_whenOtherFieldIsNotNull_isTrue() {
        ObjectToValidate object = new ObjectToValidate(null, "field2");

        Set<ConstraintViolation<ObjectToValidate>> actualViolations = validator.validate(object);

        assertThat(actualViolations).isEmpty();
    }

    @Test
    public void isValid_whenBothFieldsNotNull_isTrue() {
        ObjectToValidate object = new ObjectToValidate("field1", "field2");

        Set<ConstraintViolation<ObjectToValidate>> actualViolations = validator.validate(object);

        assertThat(actualViolations).isEmpty();
    }

    @Test
    public void isValid_whenBothFieldsNull_addsConstraintViolations() {
        ObjectToValidate object = new ObjectToValidate(null, null);

        Set<ConstraintViolation<ObjectToValidate>> violations = validator.validate(object);

        List<Violation> actualViolations = violations.stream()
                .map(v -> new Violation(v.getMessage(), v.getPropertyPath().toString()))
                .toList();
        Violation[] expectedViolations = {
                new Violation("at least should not be null", "field1"),
                new Violation("at least should not be null", "field2")
        };
        assertThat(actualViolations).containsExactlyInAnyOrder(expectedViolations);
    }

    record Violation(String message, String path) {}

    @Data
    @AllArgsConstructor
    @AtLeastOneNotNull(fields = {"field1", "field2"})
    class ObjectToValidate {
        private String field1;
        private String field2;
    }
}

``` 

Way simpler in my opinion

Comment (anrosca) [reply med-assist-backend/src/test/java/inc/evil/medassist/common/validation/AtLeastOneNotNullConstraintValidatorTest.java:28]: It also respects the principle: "Never mock something you don't own" :)

Comment (theFaustus) [reply med-assist-backend/src/test/java/inc/evil/medassist/common/validation/AtLeastOneNotNullConstraintValidatorTest.java:28]: agree

**Review by anrosca — APPROVED** [review]

**Pull request merged**